### DISCUSSION TITLE
Type annotation for unittest assertLogs() context manager

### DISCUSSION THREAD

**Discussion by jeremyn** [python/typing#1239] [General]
To make a type annotation for the context manager returned by the `unittest.TestCase.assertLogs()` method, I need to refer to the private class `_LoggingWatcher`, for example:

```python
import logging
import unittest
from unittest._log import _LoggingWatcher

def get_msgs(cm: _LoggingWatcher) -> list[str]:
    return [r.getMessage() for r in cm.records]

class Test(unittest.TestCase):
    def test(self) -> None:
        with self.assertLogs() as cm:
            logging.info("test msg")
        self.assertSequenceEqual(["test msg"], get_msgs(cm))

if __name__ == "__main__":
    unittest.main()
```
It doesn't feel right importing a private class. Is this worth creating an issue about somewhere?

**Comment by jeremyn** [comment]
Maybe I can answer my own question: I found a comment about a similar situation at https://github.com/python/typeshed/issues/2928#issuecomment-1155818548 from a couple months ago, where the poster used `typing.TYPE_CHECKING`.

Applying that comment here, I would change the first lines above to:

```python
import logging
import unittest
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from unittest._log import _LoggingWatcher

def get_msgs(cm: "_LoggingWatcher") -> list[str]:
    return [r.getMessage() for r in cm.records]
```

That seems to work. Is this the current best practice?

**Reply by srittau** [reply]
If you are uncomfortable importing the private class at runtime, this is indeed the best option at the moment. This is another case, where Python maintainers might be convinced to make the class public, though.

**Reply by jeremyn** [reply]
On further thought, I don't think there's any specific reason why they would make `unittest._log._LoggingWatcher` public, and not `hashlib._Hash` or any of the other places this might happen. However, it raises a more fundamental question of whether it's a design error to ever return a private class through a public API. Anything the caller receives is by definition public and should have a public interface, and accessing public attributes on a private class from a private module is a mixed signal at best. The library developer could `cast` a private class to a presumably more limited public one if they wanted to hide some internal details. Is this line of thought being discussed anywhere in the Python community?

In the meantime, for completeness, another approach is to create the interface on the calling side and use that, something like:

```python
from typing import Any, NamedTuple, cast

class _LogRecordInterface:
    def getMessage(self) -> str:
        return ""

class _LoggingWatcherInterface(NamedTuple):
    records: list[_LogRecordInterface]

def get_msgs(cm: Any) -> list[str]:
    return [r.getMessage() for r in cast(_LoggingWatcherInterface, cm).records]
```

**Reply by AlexWaygood** [reply]
Most interfaces in the stdlib were designed in a pre-`typing` era, when returning private classes from public APIs was less of an issue. In today's post-PEP-484 world, I'd agree with you that it's generally bad to do this.

**Reply by erictraut** [reply]
If a class is private and used only as the return type for an API, perhaps the typeshed stub should define a public protocol class (a duck type) instead of a private nominal class definition.

**Reply by jeremyn** [reply]
@erictraut Yes, I completely agree that using a protocol would have been better in my `_LoggingWatcherInterface` example above or in any place where the intent is to mock/stub a type definition.

**Reply by srittau** [reply]
Current best practice in typeshed is not to introduce public items that don't exist at runtime. I originally introduced [@typing.type_check_only](https://docs.python.org/3/library/typing.html?highlight=type_check_only#typing.type_check_only) to solve this problem, but that never took off. Maybe we should re-evaluate using this decorator in cases like this in the future.

**Reply by AlexWaygood** [reply]
> If a class is private and used only as the return type for an API, perhaps the typeshed stub should define a public protocol class (a duck type) instead of a private nominal class definition.

I'm not sure I see how this would help. It would be hard to stop the fictional protocol from drifting out of sync with the actual class returned at runtime, whereas if we're faithful to the implementation we have tools like stubtest which will warn us when the runtime changes but the stub doesn't. Moreover, if users want to import a fictional protocol, they _have_ to import it inside an `if TYPE_CHECKING` guard, whereas if we're faithful to the implementation, users don't need to.

I see this more as a problem with the implementation than a problem with typeshed.
